features/mbedtls/targets/TARGET_Silicon_Labs/crypto_aes.c

```diff
@@ -0,0 +1,545 @@
+
+#if defined(MBEDTLS_CIPHER_MODE_CFB)
+/*
+ * AES-CFB128 buffer encryption/decryption
+ */
+int mbedtls_aes_crypt_cfb128( mbedtls_aes_context *ctx,
+                              int mode,
+                              size_t length,
+                              size_t *iv_off,
+                              unsigned char iv[16],
+                              const unsigned char *input,
+                              unsigned char *output )
+{
+    size_t n = iv_off ? *iv_off : 0;
+    size_t processed = 0;
+    int ret = 0;
+    CORE_DECLARE_IRQ_STATE;
+
+    if( ctx == NULL || input == NULL || output == NULL || iv == NULL ) {
+        return ( MBEDTLS_ERR_AES_INVALID_INPUT_LENGTH );
+    }
+
+    if ( ctx->keybits != 128UL && ctx->keybits != 256UL) {
+        return MBEDTLS_ERR_AES_INVALID_KEY_LENGTH;
+    }
+
+    while ( processed < length ) {
+        if ( n > 0 ) {
+            /* start by filling up the IV */
+            if( mode == MBEDTLS_AES_ENCRYPT ) {
+                iv[n] = output[processed] = (unsigned char)( iv[n] ^ input[processed] );
+            } else {
+                int c = input[processed];
+                output[processed] = (unsigned char)( c ^ iv[n] );
+                iv[n] = (unsigned char) c;
+            }
+            n = ( n + 1 ) & 0x0F;
+            processed++;
+            continue;
+        } else {
+            /* process one ore more blocks of data */
+            CRYPTO_TypeDef *device = crypto_management_acquire();
+            device->WAC = 0;
+            device->CTRL = 0;
+
+            CORE_ENTER_CRITICAL();
+            CRYPTO_KeyBufWrite(device, (uint32_t*)ctx->key, (ctx->keybits == 128UL) ? cryptoKey128Bits : cryptoKey256Bits);
+            CRYPTO_DataWriteUnaligned(&device->DATA0, (uint8_t *)iv);
+            CORE_EXIT_CRITICAL();
+
+            /* Encryption: encrypt IV, encIV xor input -> output and IV */
+            /* Decryption: encrypt IV, encIV xor input -> output, input -> IV */
+            size_t iterations = (length - processed) / 16;
+            for (size_t i = 0; i < iterations; i++ ) {
+                device->CMD = CRYPTO_CMD_INSTR_AESENC;
+                while ((device->STATUS & CRYPTO_STATUS_INSTRRUNNING) != 0);
+
+                CORE_ENTER_CRITICAL();
+                if ( mode == MBEDTLS_AES_ENCRYPT ) {
+                    CRYPTO_DataWriteUnaligned(&device->DATA0XOR, (uint8_t *)(&input[processed]));
+                    CRYPTO_DataReadUnaligned(&device->DATA0, (uint8_t *)(&output[processed]));
+                } else {
+                    CRYPTO_DataWriteUnaligned(&device->DATA1, (uint8_t *)(&input[processed]));
+                    device->CMD = CRYPTO_CMD_INSTR_DATA1TODATA0XOR;
+                    CRYPTO_DataReadUnaligned(&device->DATA0, (uint8_t *)(&output[processed]));
+                    device->CMD = CRYPTO_CMD_INSTR_DATA1TODATA0;
+                }
+                CORE_EXIT_CRITICAL();
+                processed += 16;
```

Comment:
It can process a multiple of the block size, but then it needs to know the entire size upfront. Additionally, it cannot be preempted during the entire operation when chaining multiple blocks, and we need the preemption support (including low preemption latency) for wireless stacks.